grpc/src/main/java/io/stargate/grpc/service/ExceptionHandler.java

```diff
@@ -0,0 +1,270 @@
+package io.stargate.grpc.service;
+
+import io.grpc.Metadata;
+import io.grpc.Status;
+import io.grpc.StatusException;
+import io.grpc.StatusRuntimeException;
+import io.grpc.protobuf.ProtoUtils;
+import io.grpc.stub.StreamObserver;
+import io.stargate.proto.QueryOuterClass;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.cassandra.stargate.exceptions.AlreadyExistsException;
+import org.apache.cassandra.stargate.exceptions.CasWriteUnknownResultException;
+import org.apache.cassandra.stargate.exceptions.FunctionExecutionException;
+import org.apache.cassandra.stargate.exceptions.PersistenceException;
+import org.apache.cassandra.stargate.exceptions.ReadFailureException;
+import org.apache.cassandra.stargate.exceptions.ReadTimeoutException;
+import org.apache.cassandra.stargate.exceptions.UnavailableException;
+import org.apache.cassandra.stargate.exceptions.UnhandledClientException;
+import org.apache.cassandra.stargate.exceptions.WriteFailureException;
+import org.apache.cassandra.stargate.exceptions.WriteTimeoutException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ExceptionHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(ExceptionHandler.class);
+
+  static final Metadata.Key<QueryOuterClass.Unavailable> UNAVAILABLE_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.Unavailable.getDefaultInstance());
+  static final Metadata.Key<QueryOuterClass.WriteTimeout> WRITE_TIMEOUT_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.WriteTimeout.getDefaultInstance());
+  static final Metadata.Key<QueryOuterClass.ReadTimeout> READ_TIMEOUT_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.ReadTimeout.getDefaultInstance());
+  static final Metadata.Key<QueryOuterClass.ReadFailure> READ_FAILURE_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.ReadFailure.getDefaultInstance());
+  static final Metadata.Key<QueryOuterClass.FunctionFailure> FUNCTION_FAILURE_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.FunctionFailure.getDefaultInstance());
+  static final Metadata.Key<QueryOuterClass.WriteFailure> WRITE_FAILURE_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.WriteFailure.getDefaultInstance());
+  static final Metadata.Key<QueryOuterClass.AlreadyExists> ALREADY_EXISTS_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.AlreadyExists.getDefaultInstance());
+  static final Metadata.Key<QueryOuterClass.CasWriteUnknown> CAS_WRITE_UNKNOWN_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.CasWriteUnknown.getDefaultInstance());
+
+  private final StreamObserver<QueryOuterClass.Response> responseObserver;
+  private final AtomicBoolean exceptionOccurred = new AtomicBoolean(false);
+
+  public ExceptionHandler(StreamObserver<QueryOuterClass.Response> responseObserver) {
+
+    this.responseObserver = responseObserver;
+  }
+
+  protected void handleException(Throwable throwable) {
```

Comment:
can you add some inline comments explaining this if a bit?

Reply:
done in: https://github.com/stargate/stargate/pull/1469/commits/11f43ae296ac6db8dc1c989a1c2a7c9cdc947af9